File: external-resource-transaction/src/main/java/io/github/hyeonic/externalresourcetransaction/member/MemberService.java
package io.github.hyeonic.externalresourcetransaction.member;

import io.github.hyeonic.externalresourcetransaction.honor.Honor;
import io.github.hyeonic.externalresourcetransaction.honor.HonorRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Service
public class MemberService {

    private static final String NEWBIE = "뉴비";

    private final MemberRepository memberRepository;
    private final HonorRepository honorRepository;

    @Transactional
    public Member save(final Member member) {
        Member savedMember = memberRepository.save(member);
        honorRepository.save(new Honor(savedMember, NEWBIE));
        return savedMember;
    }
}

File: jpa-auditing/src/main/kotlin/hyeoni/c/jpaauditing/domain/MemberService.kt
package hyeoni.c.jpaauditing.domain

import hyeoni.c.jpaauditing.common.RequestContext
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class MemberService(
    private val memberRepository: MemberRepository
) {

    @Transactional
    fun save(member: Member): Member {
        RequestContext.currentAuditor = member.name
        return memberRepository.save(member)
    }

    fun getOne(id: Long): Member {
        return memberRepository.findById(id)
            .orElseThrow { IllegalArgumentException() }
    }
}
